src/libraries/Common/src/Roslyn/SyntaxValueProvider_ForAttributeWithMetadataName.cs

```diff
@@ -0,0 +1,188 @@
+        Attributes = attributes;
+    }
+}
+
+internal static partial class SyntaxValueProviderExtensions
+{
+    private static readonly char[] s_nestedTypeNameSeparators = new char[] { '+' };
+
+#if false
+
+    // Deviation from roslyn.  We do not support attributes that are nested or generic.  That's ok as that's not a
+    // scenario that ever arises in our generators.
+
+    private static readonly SymbolDisplayFormat s_metadataDisplayFormat =
+        SymbolDisplayFormat.QualifiedNameArityFormat.AddCompilerInternalOptions(SymbolDisplayCompilerInternalOptions.UsePlusForNestedTypes);
+
+#endif
+
+    /// <summary>
+    /// Creates an <see cref="IncrementalValuesProvider{T}"/> that can provide a transform over all <see
+    /// cref="SyntaxNode"/>s if that node has an attribute on it that binds to a <see cref="INamedTypeSymbol"/> with the
+    /// same fully-qualified metadata as the provided <paramref name="fullyQualifiedMetadataName"/>. <paramref
+    /// name="fullyQualifiedMetadataName"/> should be the fully-qualified, metadata name of the attribute, including the
+    /// <c>Attribute</c> suffix.  For example <c>"System.CLSCompliantAttribute</c> for <see
+    /// cref="System.CLSCompliantAttribute"/>.
+    /// </summary>
+    /// <param name="predicate">A function that determines if the given <see cref="SyntaxNode"/> attribute target (<see
+    /// cref="GeneratorAttributeSyntaxContext.TargetNode"/>) should be transformed.  Nodes that do not pass this
+    /// predicate will not have their attributes looked at at all.</param>
+    /// <param name="transform">A function that performs the transform. This will only be passed nodes that return <see
+    /// langword="true"/> for <paramref name="predicate"/> and which have a matching <see cref="AttributeData"/> whose
+    /// <see cref="AttributeData.AttributeClass"/> has the same fully qualified, metadata name as <paramref
+    /// name="fullyQualifiedMetadataName"/>.</param>
+    public static IncrementalValuesProvider<T> ForAttributeWithMetadataName<T>(
+        this SyntaxValueProvider @this,
+        IncrementalGeneratorInitializationContext context,
+        string fullyQualifiedMetadataName,
+        Func<SyntaxNode, CancellationToken, bool> predicate,
+        Func<GeneratorAttributeSyntaxContext, CancellationToken, T> transform)
+    {
+        var metadataName = fullyQualifiedMetadataName.Contains('+')
+            ? MetadataTypeName.FromFullName(fullyQualifiedMetadataName.Split(s_nestedTypeNameSeparators).Last())
+            : MetadataTypeName.FromFullName(fullyQualifiedMetadataName);
+
+        var nodesWithAttributesMatchingSimpleName = @this.ForAttributeWithSimpleName(metadataName.UnmangledTypeName, predicate);
+
+        var collectedNodes = nodesWithAttributesMatchingSimpleName
+            .Collect()
+            .WithComparer(ImmutableArrayValueComparer<SyntaxNode>.Instance)
+            /*.WithTrackingName("collectedNodes_ForAttributeWithMetadataName")*/;
```

Comment:
WithTrackingName doesn't seem to be available yet here.  so i commented them all out.